File: dp/other/robber/house_robber2.go
/*
 * @Author: yangxcc
 * @version: 1.0
 * @Date: 2022-12-09 10:56:28
 * @LastEditTime: 2022-12-09 15:46:31
 */
package robber

/**
你是一个专业的小偷，计划偷窃沿街的房屋，每间房内都藏有一定的现金。这个地方所有的房屋都 围成一圈 ，
这意味着第一个房屋和最后一个房屋是紧挨着的。同时，相邻的房屋装有相互连通的防盗系统，
如果两间相邻的房屋在同一晚上被小偷闯入，系统会自动报警 。
*/
// 可以看到，与打家劫舍问题1不同之处在于 所有的房屋都围成了一圈，其实就是最后一间房是与第一间房相邻的
func rob2(nums []int) int {
	// 在这道题目中根据题意，只有三种情况：第一间房被抢最后一间房不被抢，第一间房不被抢最后一间房被抢，或者是第一和最后都不被抢
	// 很明显，第一间和最后一间不被抢是最小的，直接忽略即可
	if len(nums) == 1 {
		return nums[0]
	}
	return max(rob1Better(nums[1:]), rob1Better(nums[:len(nums)-1]))
}

func rob2WithRecursion(nums []int) int {
	if len(nums) == 1 {
		return nums[0]
	}
	// 在[start:end]范围内，能够盗窃到的最大值
	var process func(nums []int, start, end int) int
	process = func(nums []int, start, end int) int {
		if start > end {
			return 0
		}
		if start == end {
			return nums[start]
		}

		return max(nums[start]+process(nums, start+2, end), process(nums, start+1, end))
	}

	return max(process(nums, 0, len(nums)-2), process(nums, 1, len(nums)-1))
}

File: dp/other/robber/HouseRobber2.java
/*
 * @Author: yangxcc
 * @version: 1.0
 * @Date: 2022-12-09 10:56:38
 * @LastEditTime: 2023-02-07 21:07:49
 */
package dp.other.robber;

/**
 * leetcode 213 middle 打家劫舍2
 * 
 * 你是一个专业的小偷，计划偷窃沿街的房屋，每间房内都藏有一定的现金。
 * 这个地方所有的房屋都 围成一圈 ，这意味着第一个房屋和最后一个房屋是紧挨着的。
 * 同时，相邻的房屋装有相互连通的防盗系统，如果两间相邻的房屋在同一晚上被小偷闯入，系统会自动报警 。
 * 
 * 给定一个代表每个房屋存放金额的非负整数数组，计算你 在不触动警报装置的情况下 ，今晚能够偷窃到的最高金额。
 */
public class HouseRobber2 {
    // 头尾相连(环形)
    public int rob(int[] nums) {
        if (nums.length == 1) {
            return nums[0];
        }
        return Math.max(robHelper(nums, 0, nums.length - 2), robHelper(nums, 1, nums.length - 1));
    }

    /**
     * 因为头尾相连的话，只有三种情况，头尾都不偷，头偷尾不偷，头不偷尾偷，第一种情况肯定是最少的，直接不考虑
     * 
     * @param nums
     * @param start
     * @param end
     * @return
     */
    public int robHelper(int[] nums, int start, int end) {
        int n = nums.length;
        // dp[i]表示的是从第i间房开始偷的最大收益
        int[] dp = new int[n + 2];

        for (int i = end; i >= start; i--) {
            dp[i] = Math.max(dp[i + 1], dp[i + 2] + nums[i]);
        }

        return dp[start];
    }
}
